lib/active_model/serializer/association.rb

```diff
@@ -9,10 +9,58 @@ class Serializer
     # @example
     #  Association.new(:comments, CommentSummarySerializer)
     #
-    Association = Struct.new(:name, :serializer, :options, :links, :meta) do
+    Association = Struct.new(:name, :serializer, :options) do
+      attr_reader :reflection_options, :object
+
+      def initialize(*)
+        super
+        @reflection_options = options.delete(:new_options)
+      end
+
       # @return [Symbol]
       def key
-        options.fetch(:key, name)
+        @key ||= options.fetch(:key) { name }
+      end
+
+      def links
```

Comment:
In my wip sketch of how associations work, you see everything else comes from Field

```
ActiveModel::Serializer # serializer_instance
|>  #as_json(adapter_opts = nil)
    |>  #serializable_hash(adapter_options, options, adapter_instance)
        adapter_options ||= {}
        include_directive = options.fetch(:include_directive) { ActiveModel::Serializer.include_directive_from_options(adapter_options) }
        fields = options[:fields]
        |
        |>  #attributes(fields)
        |   self.class._attributes_data.each_with_object({}) do |(key, attr), hash|
        |     |>(attr) ActiveModel::Serializer::Attribute (is a ActiveModel::Serializer::Field)
        |       next if attr.excluded?(self) # serializer_instance
        |       next unless fields.nil? || fields.include?(key)
        |       hash[key] = attr.value(self) # read_attribute_for_serialization # serializer_instance
        |
        |>  #associations(include_directive)
            relationships = {}
            self.class._reflections.values.each do |reflection|
            |>(reflection) is a ActiveModel::Serializer::Reflection subclass (is a ActiveModel::Serializer::Field subclass)
              next if reflection.excluded?(self) # serializer_instance
              key = reflection.options.fetch(:key, reflection.name)
              next unless include_directive.key?(key) # serializer_instance
              #reflection.build_association(self, instance_options) # serializer_instance
                association_serializer.new(
                  read_attribute_for_serialization(name) # association name
                )
                |>(association) is ActiveModel::Serializer::Association (is a struct)
                  yield |association|
                    adapter_opts = adapter_options.merge(include_directive: include_directive[association.key])
                    relationships[association.key] ||= association.serializer.serializable_hash(adapter_opts)

JSON API Adapter
  initialize
    @include_directive = JSONAPI::IncludeDirective.new(options[:include], allow_wildcard: true)
    @fieldset = options[:fieldset] || ActiveModel::Serializer::Fieldset.new(options.delete(:fields))
  serializable_hash
    is_collection = serializer.respond_to?(:each)
    serializers = is_collection ? serializer : [serializer]
    @primary = []
    @included = []
    @resource_identifiers = Set.new
    serializers.each { |serializer| process_resource(serializer, true) }
    serializers.each { |serializer| process_relationships(serializer, @include_directive) }
      |>(process_relationships)
        serializer.associations(include_directive).each do |association|
          |>(process_relationship)
            association_serializer = association.serializer
            association_include_directive = include_directive[association.key]
            if serializer.respond_to?(:each)
              serializer.each { |s| process_relationship(s, include_directive) }
              return
            end
            return unless serializer && serializer.object
            return unless process_resource(serializer, false)
            |>(process_resource)
              resource_identifier = ResourceIdentifier.new(serializer, instance_options).as_json
              return false unless @resource_identifiers.add?(resource_identifier)

              resource_object = resource_object_for(serializer)
              |>(resource_object_for)
                resource_object = ResourceIdentifier.new(serializer, instance_options).as_json
                requested_fields = fieldset && fieldset.fields_for(resource_object[:type])
                attributes = attributes_for(serializer, requested_fields)
                |>(attributes_for)
                |
                |>(relationships_for)
                  include_directive = JSONAPI::IncludeDirective.new(
                    fieldset.fields_for(resource_object[:type]) || '*'
                    requested_associations,
                    allow_wildcard: true
                  )
                  serializer.associations(include_directive).each_with_object({}) do |association, hash|
                    hash[association.key] = Relationship.new(args).as_json
                    |>
                      association_options = args.fetch(:options, {})
                      data =
                        if association.to_many?
                          serializer = association.serializer
                          serializer.map { |s| ResourceIdentifier.new(s, serializable_resource_options).as_json }
                        elsif association_options[:virtual_value]
                          association_options[:virtual_value]
                        elsif association.belongs_to?
                          foreign_key = association.reflection.foreign_key
                          if @object.respond_to?(foreign_key)
                            {
                              id: @object.send(foreign_key),
                              type: association.reflection.type.to_s
                            }
                          else
                            serializer && serializer.object &&
                              ResourceIdentifier.new(serializer, serializable_resource_options).as_json
                          end

                      json[:data] = data if association_options[:include_data]
```